src/Swarm/TUI/Controller.hs

```diff
@@ -450,21 +444,27 @@ validateREPLForm s =
   validate = setFieldValid (isRight result) REPLInput
 
 -- | Update our current position in the REPL history.
-adjReplHistIndex :: (Int -> Int -> Int) -> AppState -> AppState
-adjReplHistIndex (+/-) s =
-  s & uiState . uiReplHistIdx .~ newIndex
-    & (if curIndex == -1 then saveLastEntry else id)
-    & (if newIndex /= curIndex then uiState . uiReplForm %~ updateFormState newEntry else id)
+adjReplHistIndex :: TimeDir -> AppState -> AppState
+adjReplHistIndex d s =
+  ns
+    & (if s ^. replIx == s ^. replLen then saveLastEntry else id)
+    & (if s ^. replIx /= ns ^. replIx then showNewEntry else id)
```

Comment:
This line should probably express that a new entry should be shown when the entry currently focused in the REPL history changes. If not, then nothing should happen. 
Here we use a REPL index as a kind of proxy. However, wouldn't the following be clearer and even semantically more correct?
```suggestion
    & (if oldEntry /= newEntry then showNewEntry else id)
```
Here `oldEntry` can be defined in analogy to `newEntry` as follows:
```haskell
oldEntry = s ^. repl . to (fromMaybe replLast . getCurrentItemText)
```